engine.py

```diff
@@ -258,17 +283,33 @@ def safe_dialog_exec(self, func):
 
         # Merge operation can cause max dialogs to pop up, and closing the window results in a crash.
         # So keep alive and hide all of our qt windows while this type of operations are occuring.
-        for dialog in self._safe_dialog:
-            dialog.hide()
-            dialog.lower()
+        from sgtk.platform.qt import QtGui
+        toggle_map = dict()
 
-            func()
+        for dialog in self._safe_dialog:
+            needs_toggling = dialog.isVisible()
+            toggle_map[dialog] = needs_toggling
```

Comment:
nitpick: could you just build a list of dialogs that need toggling? i.e. is the `dict` necessary?